csrc/preseg_passes/fmin_fmax_promotion.cpp

```diff
@@ -0,0 +1,298 @@
+// min/max promotion.
+//
+// "BAD" statuses indicate ID's that potentially lost their NANs.
+// "GOOD" statuses will repair BAD statuses.
+// "REDUCE" statuses are reduced dimensions, and should never be mapped to ID's
+//   with non-reduced status.
+// All other statuses are "full-sized" (besides NONE which says nothing).
+//
+// A reduction ID will convert a full-size status into a REDUCE status. Likewise
+// a broadcast ID will convert a reduced status to a full size status.
+//
+// Statuses can interact via binary ops in a kind of algebra. The two reduced
+// statuses have a simple interaction: GOOD_REDUCE beats BAD-REDUCE.
+//
+// The 4 full-sized statuses have a slightly more complicated interaction:
+// 1. Any status matched with itself, produces itself.
+// 2. GOOD_BROADCAST matched with anything produces GOOD_BROADCAST.
+// 3. All other cases produce BAD_BROADCAST_DEFAULT
+//
+// An example of what each status means is below (with *max* reduction):
+// [0.0  1.0  2.0  3.0  NAN  5.0] <- DEFAULT
+// [5.0]                          <- BAD_REDUCE
+// [NAN]                          <- GOOD_REDUCE
+// [5.0  5.0  5.0  5.0  5.0  5.0] <- BAD_BROADCAST
+// [5.0  5.0  5.0  5.0  NAN  5.0] <- BAD_BROADCAST_DEFAULT
+// [NAN  NAN  NAN  NAN  NAN  NAN] <- GOOD_BROADCAST
+enum class IterDomainStatus {
+  // NONE is the status when hitting untracked ID's.
+  NONE,
+
+  // Reduced statuses
+  BAD_REDUCE,
+  GOOD_REDUCE,
+
+  // Full-size statuses
+  DEFAULT,
+  BAD_BROADCAST,
+  BAD_BROADCAST_DEFAULT,
+  GOOD_BROADCAST,
+};
+
+using IterStatusMap = std::unordered_map<IterDomain*, IterDomainStatus>;
+
+IterDomainStatus BopStatus(IterDomainStatus lhs, IterDomainStatus rhs) {
+  if (lhs == IterDomainStatus::NONE) {
+    return rhs;
+  }
+
+  if (rhs == IterDomainStatus::NONE) {
+    return lhs;
+  }
+
+  if (lhs == IterDomainStatus::GOOD_REDUCE ||
+      rhs == IterDomainStatus::GOOD_REDUCE) {
+    return IterDomainStatus::GOOD_REDUCE;
+  } else if (
+      lhs == IterDomainStatus::BAD_REDUCE &&
+      rhs == IterDomainStatus::BAD_REDUCE) {
+    return IterDomainStatus::BAD_REDUCE;
+  }
+
+  if (lhs == rhs) {
+    return lhs;
+  }
+
+  if (lhs == IterDomainStatus::GOOD_BROADCAST ||
+      rhs == IterDomainStatus::GOOD_BROADCAST) {
+    return IterDomainStatus::GOOD_BROADCAST;
+  }
+
+  // The only remaining cases are combinations of DEFAULT,
+  // BAD_BROADCAST, and BAD_BROADCAST_DEFAULT.
+  return IterDomainStatus::BAD_BROADCAST_DEFAULT;
+}
+
+bool StatusIsBad(IterDomainStatus status) {
+  return status == IterDomainStatus::BAD_REDUCE ||
+      status == IterDomainStatus::BAD_BROADCAST ||
+      status == IterDomainStatus::BAD_BROADCAST_DEFAULT;
+}
+
+bool AnyBadInputs(Expr* expr, IterStatusMap& iterMap) {
+  for (auto input : expr->inputs()) {
+    if (auto* in_tv = dynamic_cast<TensorView*>(input)) {
+      for (IterDomain* id : in_tv->getLogicalDomain()) {
+        IterDomainStatus status = iterMap[id];
```

Comment:
Is `iterMap` guaranteed to have a mapping for `id`? If so, let's use `at` so that we can mark `iterMap` as a const ref.